libpimeval/src/pimCmd.cpp

```diff
@@ -614,70 +616,79 @@ pimCmdRedSum<T>::execute()
     return false;
   }
 
-  if (pimSim::get()->getDeviceType() != PIM_FUNCTIONAL) {
-    pimObjInfo &objSrc = m_device->getResMgr()->getObjInfo(m_src);
-    objSrc.syncFromSimulatedMem();
-  }
-
   const pimObjInfo& objSrc = m_device->getResMgr()->getObjInfo(m_src);
   unsigned numRegions = objSrc.getRegions().size();
 
   // prepare per-region storage
-  m_regionSum.resize(numRegions, 0);
+  m_regionResult.resize(numRegions, 0);
 
   computeAllRegions(numRegions);
 
   // reduction
-  for (unsigned i = 0; i < numRegions; ++i) {
-    *m_result += m_regionSum[i];
+  if (m_cmdType == PimCmdEnum::REDSUM || m_cmdType == PimCmdEnum::REDSUM_RANGE) {
+    *m_result = std::accumulate(m_regionResult.begin(), m_regionResult.end(), T(0));
+  } else if (m_cmdType == PimCmdEnum::REDMIN || m_cmdType == PimCmdEnum::REDMIN_RANGE) {
+    *m_result = *std::min_element(m_regionResult.begin(), m_regionResult.end());
+  } else if (m_cmdType == PimCmdEnum::REDMAX || m_cmdType == PimCmdEnum::REDMAX_RANGE) {
+    *m_result = *std::max_element(m_regionResult.begin(), m_regionResult.end());
   }
 
   updateStats();
   return true;
 }
 
-//! @brief  PIM CMD: redsum non-ranged/ranged - sanity check
-template <typename T> bool
-pimCmdRedSum<T>::sanityCheck() const
-{
-  pimResMgr* resMgr = m_device->getResMgr();
-  if (!isValidObjId(resMgr, m_src) || !m_result) {
-    return false;
-  }
-  return true;
-}
-
-//! @brief  PIM CMD: redsum non-ranged/ranged - compute region
 template <typename T> bool
-pimCmdRedSum<T>::computeRegion(unsigned index)
+pimCmdReduction<T>::computeRegion(unsigned index)
 {
   const pimObjInfo& objSrc = m_device->getResMgr()->getObjInfo(m_src);
+  bool isVLayout = objSrc.isVLayout();
+  unsigned bitsPerElement = objSrc.getBitsPerElement();
+
   const pimRegion& srcRegion = objSrc.getRegions()[index];
+  PimCoreId coreId = srcRegion.getCoreId();
+  pimCore& core = m_device->getCore(coreId);
 
   unsigned numElementsInRegion = srcRegion.getNumElemInRegion();
   uint64_t currIdx = srcRegion.getElemIdxBegin();
+  
+  if (m_cmdType == PimCmdEnum::REDSUM || m_cmdType == PimCmdEnum::REDSUM_RANGE) {
+    m_regionResult[index] = 0;
+  } else if (m_cmdType == PimCmdEnum::REDMIN || m_cmdType == PimCmdEnum::REDMIN_RANGE) {
+    m_regionResult[index] = std::numeric_limits<T>::max();
+  } else if (m_cmdType == PimCmdEnum::REDMAX || m_cmdType == PimCmdEnum::REDMAX_RANGE) {
+    m_regionResult[index] = std::numeric_limits<T>::lowest();
+  }
+
   for (unsigned j = 0; j < numElementsInRegion && currIdx < m_idxEnd; ++j) {
     if (currIdx >= m_idxBegin) {
-      uint64_t operandBits = objSrc.getElementBits(currIdx);
-      T operand = pimUtils::signExt(operandBits, objSrc.getDataType());
-      m_regionSum[index] += operand;
+      auto locSrc = srcRegion.locateIthElemInRegion(j);
```

Comment:
Oh I see. This code is outdated and needed to be merged with the latest main branch.